multicluster/service-mirror/cluster_watcher.go

```diff
@@ -815,3 +906,331 @@ func (rcsw *RemoteClusterServiceWatcher) createOrUpdateEndpoints(ctx context.Con
+	if err != nil {
+		return err
+	}
+	localServiceName := rcsw.mirroredResourceName(remoteService.Name)
+	serviceInfo := fmt.Sprintf("%s/%s", remoteService.Namespace, remoteService.Name)
+
+	// To mirror a headless service, we need to create a clusterIP for each
+	// host in its Endpoints.
+	hosts := make(map[string]struct{})
+	subsetsToCreate := make([]corev1.EndpointSubset, 0, len(remoteEp.Subsets))
+	for _, subset := range remoteEp.Subsets {
+		newAddresses := make([]corev1.EndpointAddress, 0, len(subset.Addresses))
+		for _, addr := range subset.Addresses {
+			// Create annotations for host's service
+			if addr.TargetRef == nil {
+				// ignore addresses w/o a backing pod (if they exist)
+				continue
+			}
+
+			refLocalName := rcsw.mirroredResourceName(addr.TargetRef.Name)
+			createdService, err := rcsw.createHeadlessMirroredHost(ctx, addr.TargetRef, remoteService, refLocalName)
+			if err != nil {
+				rcsw.log.Errorf("failed to create headless mirrored hosts for %s: %v", refLocalName, err)
+				continue
+			}
+
+			hosts[addr.TargetRef.Name] = struct{}{}
+			newAddresses = append(newAddresses, corev1.EndpointAddress{
+				Hostname: addr.TargetRef.Name,
+				IP:       createdService.Spec.ClusterIP,
+			})
+
+		}
+
+		if len(newAddresses) == 0 {
+			continue
+		}
+
+		subsetsToCreate = append(subsetsToCreate, corev1.EndpointSubset{
+			Addresses: newAddresses,
+			Ports:     subset.DeepCopy().Ports,
+		})
+	}
+
+	endpointsToCreate := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      localServiceName,
+			Namespace: remoteService.Namespace,
+			Labels: map[string]string{
+				consts.MirroredResourceLabel:  "true",
+				consts.RemoteClusterNameLabel: rcsw.link.TargetClusterName,
+			},
+			Annotations: map[string]string{
+				consts.RemoteServiceFqName: fmt.Sprintf("%s.%s.svc.%s", remoteService.Name, remoteService.Namespace, rcsw.link.TargetClusterDomain),
+			},
+		},
+		Subsets: subsetsToCreate,
+	}
+
+	if rcsw.link.GatewayIdentity != "" {
+		endpointsToCreate.Annotations[consts.RemoteGatewayIdentity] = rcsw.link.GatewayIdentity
+	}
+
+	rcsw.log.Infof("Creating a new root Endpoints for %s", serviceInfo)
+	if _, err := rcsw.localAPIClient.Client.CoreV1().Endpoints(remoteService.Namespace).Create(ctx, endpointsToCreate, metav1.CreateOptions{}); err != nil {
+		// we clean up after ourselves
+		rcsw.localAPIClient.Client.CoreV1().Services(remoteService.Namespace).Delete(ctx, localServiceName, metav1.DeleteOptions{})
+		// and retry
+		return RetryableError{[]error{err}}
+	}
+
+	return nil
+}
+
+func (rcsw *RemoteClusterServiceWatcher) deleteHeadlessMirrors(namespace, name string) error {
+	ep, err := rcsw.localAPIClient.Endpoint().Lister().Endpoints(namespace).Get(name)
+	if err != nil {
+		return err
+	}
+
+	if err = checkEndpointsValid(ep); err != nil {
+		return err
+	}
+
+	for _, subset := range ep.Subsets {
+		for _, address := range subset.Addresses {
+			if address.Hostname == "" {
```

Comment:
should it be possible to hit this?  presumably we only call `deleteHeadlessMirrors` on a headless mirror service, all of who's endpoints should have hostnames.

Reply:
You're right, it shouldn't be possible, unless they were manually deleted -- I don't find it to be a very plausible case though so I'm okay with removing this.